src/main/java/org/jenkinsci/maven/plugins/hpi/TestDependencyMojo.java

```diff
@@ -52,5 +276,410 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+            }
+        } catch (IOException e) {
+            throw new MojoExecutionException("Failed to scan " + war, e);
+        }
+        return overrides;
+    }
+
+    private static Map<String, String> parseOverrides(List<String> overrideVersions) throws MojoExecutionException {
+        Map<String, String> overrides = new HashMap<>();
+        for (String override : overrideVersions) {
+            Matcher m = Pattern.compile(OVERRIDE_REGEX).matcher(override);
+            if (!m.matches()) {
+                throw new MojoExecutionException("illegal override: " + override);
+            }
+            overrides.put(m.group(1), m.group(2));
+        }
+        return overrides;
+    }
+
+    /**
+     * Apply the overrides specified by the user or upper bounds analysis to the model (i.e.,
+     * dependency management or dependencies) in the shadow project. This clears the existing
+     * resolution that was done because of the {@code @requiresDependencyResolution} Mojo attribute,
+     * as it is now invalid. It is possible to perform such a pass manually on a plugin and compare
+     * the results with this algorithm to verify that the logic in this method is correct.
+     */
+    private static void applyOverrides(
+            Map<String, String> overrides,
+            Map<String, String> bundledPlugins,
+            MavenProject project,
+            Log log)
+            throws MojoExecutionException {
+        Set<String> appliedOverrides = new HashSet<>();
+        Set<String> appliedBundledPlugins = new HashSet<>();
+
+        // Update existing dependency entries in the model.
+        for (Dependency dependency : project.getDependencies()) {
+            String key = toKey(dependency);
+            if (updateDependency(dependency, overrides, "direct dependency", log)) {
+                appliedOverrides.add(key);
+            }
+            if (updateDependency(dependency, bundledPlugins, "direct dependency", log)) {
+                appliedBundledPlugins.add(key);
+            }
+        }
+
+        // Update existing dependency management entries in the model.
+        if (project.getDependencyManagement() != null) {
+            for (Dependency dependency : project.getDependencyManagement().getDependencies()) {
+                String key = toKey(dependency);
+                if (updateDependency(dependency, overrides, "dependency management entry", log)) {
+                    appliedOverrides.add(key);
+                }
+                if (updateDependency(dependency, bundledPlugins, "dependency management entry", log)) {
+                    appliedBundledPlugins.add(key);
+                }
+            }
+        }
+
+        /*
+         * If an override was requested for a transitive dependency that is not in the model, add a
+         * dependency management entry to the model.
```

Comment:
Yeah, that and the application of upper bounds analysis. While it might be technically feasible to munge the XML file from the POM-rewriting part of PCT with the right algorithm, it isn't possible to determine upper bounds without actual dependency resolution.